packages/middleware-joi/src/joi.middleware.ts

```diff
@@ -0,0 +1,34 @@
+import * as Joi from 'joi';
+import { Schema, SchemaValidator } from './schema';
+import { Effect, HttpRequest, HttpError, HttpStatus } from '@marblejs/core';
+import { from, of, throwError } from 'rxjs';
+import { mergeMap, flatMap, catchError, mapTo, tap, switchMap, toArray, map } from 'rxjs/operators';
+
+const validateSource = (req: HttpRequest, rules: Map<string, any>, options) => from(rules.keys()).pipe(
+  mergeMap(rule => of(req[rule]).pipe(
+    flatMap(item => from(Joi.validate(item || {}, rules.get(rule), options))),
+    catchError((err: Error) => throwError(new HttpError(err.message, HttpStatus.BAD_REQUEST))),
+    map(result => (req[rule] = result)),
+  )),
+  toArray(),
+  mapTo(req),
+);
+
+const validator$ = (schema: Schema, options: Joi.ValidationOptions = {}): Effect<HttpRequest> => request$ => {
+  const result = Joi.validate(schema, SchemaValidator);
+  const rules = Object.keys(schema).reduce((acc, value) => acc.set(value, Joi.compile(schema[value])), new Map());
+
+  if (result && result.error) {
+    return throwError(result.error);
+  }
+
+  return request$.pipe(
+    tap(req => req.params = req.route ? req.route.params : {}),
```

Comment:
Do you think that it is possible to handle nested keys like we currently have in our `HttpRoute` interface?
I'm asking because I would like to avoid redundant parameters in `request` object. If it is hard to implement I think this is the best time to think about flattening `params` and `queries` that are currently inside `HttpRoute` object type. It will be much easier to access and we will be compatible with other *node* http libraries/frameworks like *Express*.

Reply:
@pdomaleczny @sebastianmusial Can you also join to the conversation here?

Reply:
IMO, flattening will be great because the code will be easier to use. 
I will do PR with this changes.

Reply:
@luciorubeens see: #19

Reply:
@luciorubeens merged, feel free to update 💪